File: AIiOS/AIiOS/Services/StorageService.swift
import Foundation
import FirebaseStorage
import UIKit

final class StorageService {
    static let shared = StorageService()

    private let storage = Storage.storage()
    private let maxBytes: Int64 = 4 * 1024 * 1024

    private init() {}

    func uploadAvatar(uid: String, image: UIImage) async throws -> String {
        guard let data = image.jpegData(compressionQuality: 0.8) else {
            throw UploadError.encodingFailed
        }
        guard Int64(data.count) <= maxBytes else {
            throw UploadError.imageTooLarge
        }
        let ref = storage.reference().child("avatars/\(uid).jpg")
        let meta = StorageMetadata()
        meta.contentType = "image/jpeg"
        _ = try await ref.putDataAsync(data, metadata: meta)
        let url = try await ref.downloadURL()
        return url.absoluteString
    }

    enum UploadError: LocalizedError {
        case encodingFailed, imageTooLarge
        var errorDescription: String? {
            switch self {
            case .encodingFailed: return "Could not encode image."
            case .imageTooLarge:  return "Image must be under 4 MB."
            }
        }
    }
}

File: AIAndroid/app/src/main/kotlin/ai/ecoinference/app/services/StorageService.kt
package ai.ecoinference.app.services

import android.graphics.Bitmap
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import kotlinx.coroutines.tasks.await
import java.io.ByteArrayOutputStream

object StorageService {

    private val storage = FirebaseStorage.getInstance()
    private const val MAX_BYTES = 4 * 1024 * 1024L

    suspend fun uploadAvatar(uid: String, bitmap: Bitmap): String {
        val out = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 80, out)
        val bytes = out.toByteArray()
        if (bytes.size > MAX_BYTES) throw IllegalArgumentException("Image must be under 4 MB.")

        val ref  = storage.reference.child("avatars/$uid.jpg")
        val meta = StorageMetadata.Builder().setContentType("image/jpeg").build()
        ref.putBytes(bytes, meta).await()
        return ref.downloadUrl.await().toString()
    }
}
